### PR TITLE
Use IntegralRange in fgOptimizeRelationalComparisonWithCasts

### LINKED ISSUE #69200: JIT perf: Span.Slice(0, ...) emits unnecessary 64-bit extensions
When calling `Span<T>.Slice(offset, length)`, the internal implementation 64-bit extends _offset_ and _length_, then compares their sum against _span.Length_. This is an optimization that allows us to get away with a single comparison.

```cs
public class C {
    public static void M(ReadOnlySpan<byte> span, int offset, int length) {
        DoIt(span.Slice(offset, length));
    }
    
    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void DoIt(ReadOnlySpan<byte> span)
    {
    }
}
```
```asm
C.M(System.ReadOnlySpan`1<Byte>, Int32, Int32)
    L0000: sub rsp, 0x38
    L0004: xor eax, eax
    L0006: mov [rsp+0x28], rax
    L000b: mov eax, edx  ; rax := 64-bit zero-extension of offset
    L000d: mov edx, r8d  ; rdx := 64-bit zero-extension of length 
    L0010: add rdx, rax  ; rdx := sum of offset and length
    L0013: mov r9d, [rcx+8]  ; r9 := 64-bit zero-extension of span.Length
    L0017: cmp rdx, r9  ; bounds check
    L001a: ja short L003c
    L001c: mov rcx, [rcx]
    L001f: add rcx, rax
    L0022: mov [rsp+0x28], rcx
    L0027: mov [rsp+0x30], r8d
    L002c: lea rcx, [rsp+0x28]
    L0031: call 0x00007ffbf1df0020
    L0036: nop
    L0037: add rsp, 0x38
    L003b: ret
    L003c: call 0x00007ffbe86e5778
    L0041: int3
```

However, when _start = 0_ is a known constant, JIT still emits the 64-bit extension before performing the comparison. This is unnecessary because a comparison of two zero-extended values will yield the same result as a comparison of the original values. It also burns a register.

```cs
    public static void M(ReadOnlySpan<byte> span, int length) {
        DoIt(span.Slice(0, length));
    }
```
```asm
C.M(System.ReadOnlySpan`1<Byte>, Int32)
    L0000: sub rsp, 0x38
    L0004: xor eax, eax
    L0006: mov [rsp+0x28], rax
    L000b: mov eax, edx  ; rax := unnecessary 64-bit zero-extension of length
    L000d: mov r8d, [rcx+8]  ; r8 := unnecessary 64-bit zero-extension of span.Length
    L0011: cmp rax, r8  ; bounds check
    L0014: ja short L0032
    L0016: mov rcx, [rcx]
    L0019: mov [rsp+0x28], rcx
    L001e: mov [rsp+0x30], edx
    L0022: lea rcx, [rsp+0x28]
    L0027: call 0x00007ffbf1e10020
    L002c: nop
    L002d: add rsp, 0x38
    L0031: ret
    L0032: call 0x00007ffbe86e5778
    L0037: int3
```

In the above example, I would have expected the three annotated instructions to be collapsed into the single instruction `cmp edx, dword ptr [rcx + 8]`.

### PR DESCRIPTION
**Pull request by jakobbotsch**
Fix #69200

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1773186&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by jakobbotsch** [comment]
Hmm, looks like I need to update `IntegralRange::ForNode` to do something for `GT_CNS_INT` if we want it to completely subsume this logic here.

**Comment by jakobbotsch** [comment]
Diff for example in #69200:
```diff
        mov      qword ptr [rsp+28H], rax
 						;; size=11 bbWeight=1    PerfScore 1.50
 G_M28380_IG02:
-       mov      eax, edx
-       mov      r8d, dword ptr [rcx+8]
-       cmp      rax, r8
+       cmp      edx, dword ptr [rcx+8]
        ja       SHORT G_M28380_IG04
        mov      rcx, bword ptr [rcx]
        mov      bword ptr [rsp+28H], rcx
@@ -43,7 +41,7 @@ G_M28380_IG02:
        lea      rcx, bword ptr [rsp+28H]
        call     [Program:DoIt(ReadOnlySpan`1)]
        nop      
-						;; size=35 bbWeight=1    PerfScore 11.25
+						;; size=29 bbWeight=1    PerfScore 11.75
 G_M28380_IG03:
        add      rsp, 56
        ret      
@@ -53,6 +51,6 @@ G_M28380_IG04:
        int3     
 						;; size=7 bbWeight=0    PerfScore 0.00
 
-; Total bytes of code 58, prolog size 11, PerfScore 19.80, instruction count 17, allocated bytes for code 58 (MethodHash=f8b59123) for method Program:M(ReadOnlySpan`1,int)
+; Total bytes of code 52, prolog size 11, PerfScore 19.70, instruction count 15, allocated bytes for code 52 (MethodHash=f8b59123) for method Program:M(ReadOnlySpan`1,int)
 ; ============================================================
```

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib @GrabYourPitchforks

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -13336,97 +13336,97 @@ GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* cmp)
...
-        if (!knownPositiveFitsIntoU32)
+        assert(op->OperIs(GT_CAST));
+        if (op->AsCast()->IsUnsigned())
         {
-            return cmp;
+            return true;
         }
 
+        return IntegralRange(SymbolicIntegerValue::Zero, SymbolicIntegerValue::IntMax)
+            .Contains(IntegralRange::ForNode(op->AsCast()->CastOp(), this));
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:13393]: Nit: could be `IntegralRange::ForNode(op->AsCast()->CastOp(), this).IsPositive()`.

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -13336,97 +13336,97 @@ GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* cmp)
...
-
-        if (knownPositiveOp->OperIs(GT_CAST))
-        {
-            // Drop cast from knownPositiveOp too
-            if (knownPositiveIsOp2)
+        auto transform = [this](GenTree** use) {
+            if ((*use)->IsIntegralConst())
             {
-                cmp->gtOp2 = knownPositiveOp->AsCast()->CastOp();
+                (*use)->BashToConst(static_cast<int>((*use)->AsIntConCommon()->LngValue()), TYP_INT);
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:13413]: Nit: no need to specify `TYP_INT`, the idea with `BashToConst` is that `static_cast<int>` implies it.

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -13336,97 +13336,97 @@ GenTree* Compiler::fgOptimizeRelationalComparisonWithCasts(GenTreeOp* cmp)
...
+        return IntegralRange(SymbolicIntegerValue::Zero, SymbolicIntegerValue::IntMax)
+            .Contains(IntegralRange::ForNode(op->AsCast()->CastOp(), this));
+    };
+
+    bool op1UpperIsZero = isUpperZero(op1);
+    bool op2UpperIsZero = isUpperZero(op2);
+
+    // If both operands have zero as the upper half we can optimize then any
+    // signed/unsigned 64-bit comparison is equivalent to the same unsigned
+    // 32-bit comparison.
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:13401]: Ugh, phrasing in this comment... will fix this as part of a follow-up.

**Review by SingleAccretion — APPROVED** [review]
Looks great (modulo nits)!

**Review by EgorBo — APPROVED** [review]

**Pull request merged**